src/masternodes/mn_checks.cpp

```diff
@@ -4492,6 +4492,14 @@ Res CPoolSwap::ExecuteSwap(CCustomCSView& view, std::vector<DCT_ID> poolIDs, boo
             }
         }
 
+        auto forward = swapAmount.nTokenId == pool->idTokenA;
+
+        if (height >= static_cast<uint32_t>(Params().GetConsensus().GrandCentralHeight) && lastSwap) {
+            if ((forward && pool->idTokenB != obj.idTokenTo) || (!forward && pool->idTokenA != obj.idTokenTo)) {
+                return Res::Err("Final swap output is not same as idTokenTo");
+            }
+        }
+
```

Comment:
Would it just not be cleaner to add the check at the very end on the `result` and reject it? Keeps the swap logic lighter.